std/typecons.d

```diff
@@ -1787,6 +1798,187 @@ unittest
     assert(rebindable(pr3341_aa)[321] == 543);
 }
 
+/** Models safe reassignment of otherwise constant struct instances.
+ *
+ * A struct with a field of reference type cannot be assigned to a constant
+ * struct of the same type. `Rebindable!(const S)` allows assignment to
+ * a `const S` while enforcing only constant access to its fields.
+ *
+ * `Rebindable!(immutable S)` does the same but field access may create a
+ * temporary copy of `S` in order to enforce _true immutability.
+ */
+template Rebindable(S)
+if (is(S == struct))
+{
+    static if (isMutable!S)
+        alias Rebindable = S;
+    else
+    struct Rebindable
+    {
+        // fields of mutPayload must be treated as tail const
+        private union
+        {
+            Unqual!S mutPayload = void;
+            S payload;
+        }
+
+    @trusted:
```

Comment:
I think it's OK now, opAssign is never called, and `S.~this` has its safety inferred. `Rebindable.this `also uses inferrence for safety of `S.this`.